### DISCUSSION TITLE
Engine HAT voltage measurement via ADS1115

### DISCUSSION THREAD

**Discussion by wabson** [hatlabs/discussions#56] [Q&A] [answered]
Hi, I am attempting to use the analogue inputs on the engine HAT connected to my ESP32 to measure the voltage of my boat house battery, but I am finding that the voltage returned is around 1 volt, which obviously is not right.

I have modified the [example project](https://github.com/hatlabs/SH-ESP32-engine-hat-firmware), which uses the ADS1115 to measure tank levels, and simplified the code to return the voltage readings directly, in my case, i.e.

```
new RepeatSensor<float>(ads_read_delay, [ads1115, channel]() {
        int16_t adc_output = ads1115->readADC_SingleEnded(channel);
        float adc_output_volts = ads1115->computeVolts(adc_output);
        return adc_output_volts;
      });
```

The problem I am having is that the `computeVolts()` returns a consistent 1.01 volts, rather than the 13.5-14 volts I expect to see when the engine is running. I have checked the [Adafruit library sample code](https://github.com/adafruit/Adafruit_ADS1X15/blob/master/examples/singleended/singleended.ino) (docs are pretty light) and I am guessing this is because as it notes, with a gain of 1 the ADC is able to measure between -4.96 and +4.96 volts, so presumably the engine HAT is reducing the voltage before it goes into the ADC? Apologies if I have missed it, but I could not find any notes in the [docs](https://docs.hatlabs.fi/sh-esp32/pages/add-ons/engine-hat/hardware/) around how the ADS115 deals with the up to 29V specified as valid input voltages that would help me apply any scaling factor needed.

Perhaps I've done something wrong on the hardware side and the Texas Instruments ADC does genuinely return the real voltage, but it's hard to know what if so, as I just have a single wire running to the battery (+) into the upper pin of the analogue input channel on that HAT, with the GND pin being wired to the GND from the power input. Testing with a multimeter with the board powered off seems to show the correct voltage as expected.

**Comment by mairas** [comment, marked as answer]
Your hunch is correct: there is a voltage divider to reduce the maximum voltage level. The schematics PDF is on this web page: https://docs.hatlabs.fi/sh-esp32/pages/add-ons/engine-hat/hardware/. See page 2 of the schematics, the voltage divider circuitry is in the second column from right.

The divider is set to scale the voltage by $2.048/29$. Hence, to get your battery voltage, multiply `adc_output_volts` by $29/2.048$. 

It seems like it indeed wasn't documented elsewhere. I've now added a mention to the hardware doc page.

**Reply by wabson** [reply]
Thanks @mairas that indeed makes sense and you are right, I now see that in the schematic too. Thanks for the info and for updating the docs.

**Comment by mairas** [comment]
FWIW, 1.01V raw input level results in a voltage of 14.3V which is totally plausible assuming that you have the engine or some other battery charger on.
